### PR TITLE
QAMR + QA-SRL Update

### PR DESCRIPTION
**Pull request by zphang**
* Addition of QAMR task
* Switching out QA-SRL evaluation
* Moved some span metrics over to `allennlp_mods/span_metrics.py`

Notes on QAMR:
* QAMR's data is PTB tokenize, so we first detokenize it using sacremoses. All subsequent processing, including evaluation is based on the detokenized data.
* QAMR has both a given answer (provided by the first annotator, who writes both the question and annotates the answer) as well as two validator answers (from two other annotators). We're currently using only the first-annotator answer, on the filtered data (the filtering is done based on the other two validation annotator feedback, but we're not using their provided answers). 
* Also, the QAMR answers are given as individual tokens, but the paper states that the answer spans are usually but not always contiguous. We're simply taking the min/max to get the full span currently.
* Conversion from PTB token-spans is based on `get_aligner_fn`. We also use this to map back from predicted (e.g. BERT) tokens to the original string for the purposes of SQuAD-style scoring. The downside of relying on `get_aligner_fn` is that the SQuAD-scored predictions can only go up to the granularity of space-tokenized tokens. This is usually not an issue because SQuAD scoring normalizes for spacing and punctuation.

Notes on QA-SRL:
* Similar to QAMR, we move over to full SQuAD-style scoring with space-token spans.

I also wrote some additional functions that are not currently used (e.g. for detokenizing BERT tokens to raw strings, if we do move away from `get_aligner_fn`'s reliance on space-tokenization).

### CODE REVIEW

File: jiant/models.py [hunk 1]
```diff
@@ -964,32 +964,59 @@ def _span_prediction_forward(self, batch, task, predict):
...
-            task.update_metrics(
-                logits=logits_dict,
-                labels={"span_start": batch["span_start"], "span_end": batch["span_end"]},
+            pred_span_end_i = min(
+                max(pred_span_end_i, pred_span_start_i + 1),
+                len(batch["space_processed_token_map"][i]) - 1,
+            )
+
+            # space_processed_token_map is a list of tuples
+            #   (space_token, processed_token (e.g. BERT), space_token_index)
```
Comment (pruksmhc) [inline jiant/models.py]: period after ).

Comment (pruksmhc) [reply jiant/models.py]: Also put periods after sentences below.

File: jiant/allennlp_mods/span_metrics.py [hunk 2]
```diff
@@ -1,47 +1,97 @@
-import numpy as np
-import torch
+import collections
```
Comment (pruksmhc) [inline jiant/allennlp_mods/span_metrics.py]: Can you move this to jiant/metrics?

File: jiant/tasks/qa.py [hunk 3]
```diff
@@ -608,3 +574,243 @@ def preprocess_qasrl_datum(cls, datum):
...
+
+
+@register_task("qamr", rel_path="QAMR/")
+class QAMRTask(SpanPredictionTask):
+    """ Question-Answer Meaning Representation (QAMR)
+        https://github.com/uwnlp/qamr
+    """
+
+    def __init__(self, path, max_seq_len, name="qamr", **kw):
+        """ There are 1363 supertags in CCGBank without introduced token. """
```
Comment (pruksmhc) [inline jiant/tasks/qa.py]: This doesn't belong here.

**Comment by pruksmhc** [comment]
What's the status of this?

**Comment by pruksmhc** [comment]
@zphang  is this ready for a review?
